pkg/crunch42/crunch.go

```diff
@@ -0,0 +1,226 @@
+
+	toReturn := &Item{}
+	path := fmt.Sprintf("%s/collections", c.baseURL)
+	resp, err := c.DoRequest("POST", path, collection, toReturn)
+	if err != nil {
+		return nil, resp, err
+
+	}
+	return toReturn, resp, err
+}
+
+func (c *Client) CreateAPI(api *API) (*Item, *ErrorResponse, error) {
+	i := &Item{}
+	path := "https://platform.42crunch.com/api/v2/apis"
+
+	resp, err := c.DoRequest("POST", path, api, i)
+	if err != nil {
+		return nil, resp, err
+
+	}
+	return i, resp, err
+}
+
+type Client struct {
+	client  *http.Client
+	apiKey  string
+	baseURL *url.URL
+}
+
+func NewClient(apiKey string, httpClient *http.Client) (*Client, error) {
+	if httpClient == nil {
+		httpClient = http.DefaultClient
+	}
+
+	url, err := url.Parse(baseURL)
+	if err != nil {
+		return nil, err
+	}
+	client := &Client{
+		apiKey:  apiKey,
+		baseURL: url,
+		client:  httpClient,
+	}
+
+	return client, nil
+}
+
+func (c *Client) DoRequest(method, path string, body, v interface{}) (*ErrorResponse, error) {
+	req, err := c.NewRequest(method, path, body)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.Do(req, v)
+}
+
+func (c *Client) DoFormRequestPOST(method, path string, api *API, v interface{}) (*ErrorResponse, error) {
+
+	return nil, nil
+
+}
+
+func (c *Client) NewRequest(method, path string, body interface{}) (*http.Request, error) {
+	// relative path to append to the endpoint url, no leading slash please
+	if path[0] == '/' {
+		path = path[1:]
+	}
+	rel, err := url.Parse(path)
+	if err != nil {
+		return nil, err
+	}
+
+	u := c.baseURL.ResolveReference(rel)
+	var req *http.Request
+	if body != nil {
+		bodyBytes, _ := json.Marshal(body)
+		req, _ = http.NewRequest(method, u.String(), bytes.NewBuffer(bodyBytes))
+		req.Header.Set("Content-Type", "application/json")
+	} else {
+		req, _ = http.NewRequest(method, u.String(), nil)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	req.Close = true
+
+	if _, ok := os.LookupEnv(debug); ok {
+		command, _ := http2curl.GetCurlCommand(req)
+		fmt.Println(command)
+	}
+	req.Header.Add("X-API-KEY", c.apiKey)
+	return req, nil
+}
+
+func (c *Client) Do(req *http.Request, v interface{}) (*ErrorResponse, error) {
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+	if resp.StatusCode > 299 {
+		o, _ := io.ReadAll(resp.Body)
+		errResp := &ErrorResponse{
+			StatusCode: resp.StatusCode,
+			Errors:     string(o),
+		}
+
+		return errResp, fmt.Errorf("%s returned %d", req.URL, resp.StatusCode)
+	}
+
+	o, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	if _, ok := os.LookupEnv(debug); ok {
+		fmt.Println("Response body:", string(o))
+	}
+	err = json.Unmarshal(o, v)
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, err
+}
+
+type Collection struct {
+	ID            string `json:"id"`
+	Name          string `json:"name"`
+	IsShared      bool   `json:"isShared"`
+	IsSharedWrite bool   `json:"isSharedWrite"`
+}
+
+type API struct {
```

Comment:
I'm guessing `technicalName` isn't required here because the tests pass (I've run it locally and it seems to pass without `technicalName`). I only ask because on <https://www.postman.com/get-42crunch/workspace/42crunch-api/api/0f8cfb06-3a01-418e-8191-4d5b12fd4420/request/13761657-c11df4e6-0f1e-465f-a971-df8091867f7c> they have `techincalName` specified.

Reply:
we are not using technical name anywhere

Reply:
OK.